mlir/lib/Dialect/Rock/Tuning/RockTuningImpl.cpp

```diff
@@ -55,6 +58,78 @@ void createAttnTuningRangeBF(TuningParamSet *newSpace, AttentionOp attnOp,
   }
 }
 
+SmallVector<int64_t>
+computeOptimalSplitKFactors(RockGemmWrapperInterface gemmOp,
+                            int32_t gemmMPerBlock, int32_t gemmNPerBlock,
+                            int32_t gemmKPerBlock, int32_t kPack) {
+  auto info = PopulateParamsInfo::fromOp(gemmOp);
+  SmallVector<int64_t> splitKValues = {1};
+  if (!info.gemmAType.isF32()) {
+    return splitKValues;
+  }
+
+  if (!gemmOp.getNumCU().has_value()) {
+    return splitKValues;
+  }
+  const double numCUs = gemmOp.getNumCU().value();
+
+  const InitParams params{gemmMPerBlock, gemmNPerBlock, gemmKPerBlock};
+  const GemmSize gemmSize =
+      calculatePaddedGemmSize(params, info.gemmSize, kPack);
+  const double numMTiles = gemmSize.m / gemmMPerBlock;
+  const double numNTiles = gemmSize.n / gemmNPerBlock;
+
+  auto computeImbalance = [&](int32_t splitKFactor) -> double {
+    const double totalNumWorkGroups = numMTiles * numNTiles * splitKFactor;
+    const double maxWorkGroupsPerCU = std::ceil(totalNumWorkGroups / numCUs);
+    // imbalances = max. CU work / average work per CU
+    return (maxWorkGroupsPerCU * numCUs) / totalNumWorkGroups;
+  };
+
+  const auto dataPrallelGemmImbalabce = computeImbalance(1);
+  constexpr double imbalaceThreshold = 1.20;
+  if (dataPrallelGemmImbalabce < imbalaceThreshold) {
+    return splitKValues;
+  }
+
+  struct LocalData {
+    int64_t splitKValue = 0;
+    double workImbalance = 0.0;
+  };
+  SmallVector<LocalData> factors;
+  constexpr double minGain = 1.30;
+  // There are cases where perfect load balancing can be achieved with very
+  // high splitK values. However, experiments show that performance
+  // can considerably drop in such cases. Currently, we limit the `upperBound`
+  // on purpose because the current heuristics does not consider the overheads
+  // resulting from reducing partial solution along the split dimension.
+  // This needs to be improved in the future.
+  constexpr int32_t upperBound = 32;
+  for (int32_t splitKFactor = 2; splitKFactor < upperBound; ++splitKFactor) {
+    const double imbalance = computeImbalance(splitKFactor);
+    const auto gain = dataPrallelGemmImbalabce / imbalance;
+    if (gain > minGain) {
+      factors.emplace_back(LocalData{splitKFactor, imbalance});
+    }
+  }
+
+  if (factors.empty()) {
+    return splitKValues;
+  }
+
+  llvm::sort(factors.begin(), factors.end(), [](LocalData &a, LocalData &b) {
+    return a.workImbalance < b.workImbalance;
+  });
+
+  size_t maxVariants = 6;
+  maxVariants = factors.size() > maxVariants ? maxVariants : factors.size();
```

Comment:
Do you mean who can set it? At this point nobody. It is just a hard-coded value which is serving as limiting factor for the exhaustive tuning. Otherwise, it might take too long.

This PR is only about ennoblement the splitK through the `perfConfig` string.

The heuristic needs to be improved and the ticket is here: https://github.com/ROCm/rocMLIR-internal/issues/1410

Reply:
No I mean from what I can read here maxVariants is dead code.
No one uses it ?

Reply:
Thanks for noticing it. Yep, I've reworked it based on our discussion